src/framework-ondata.js

```diff
@@ -1,18 +1,129 @@
+ * If the updater should be executed, the fRunUpdater wrapper API will be
+ * called to to handle that.  The updater output (or error) will be stored 
+ * in output or error state, and those results will be sent on to any 
+ * down stream components for their processing.
+ * 
  * @this port context
- *
- * @param payload data payload on the input port
- * @param socketIndex (optional) this will be set only for ports configured for
- *        multiple inputs (multi/noflo addressable).
+ * @param payload data coming into the port
+ * @param socketIndex socket index if the port is a multiple input (addressable) 
  */
 module.exports = function( payload, socketIndex ) {
 
+     var portName = this.name;
+
+     var vnid = payload.vnid || '';
+     var vni = this.nodeInstance.vni(vnid);
+
+     // Set the input states
+     vni.inputStates( portName, socketIndex, stateFactory( vnid, payload ) );
+
+     // check if should run updater & run if necessary
+     if ( shouldRunUpdater( this.nodeInstance, vni ) ) { 
+          
+        // Save vars we will need in the promise where the context is different
+        var outputPorts = this.nodeInstance.outPorts;
+
+        if ( _.isUndefined( this.nodeInstance.wrapper.fRunUpdater ) || 
+             ! _.isFunction( this.nodeInstance.wrapper.fRunUpdater ) ) { 
+
+             // Don't have a wrapper fRunUpdater 
+             throw Error( 'No wrapper fRunUpdater function found!  Cannot run updater.' );
+
+        } else { 
+            var promise = this.nodeInstance.wrapper.fRunUpdater.call( vni );
+            Promise.resolve( promise ).then( function( data ) { 
+
+                // Set output state and send it on over the output port
+                if (data) {
+                    var outputState = stateFactory( vnid, data );
+                    vni.outputState( outputState );
+    
+                    // Should this be in access or output layer? 
+                    outputPorts.output.send( outputState );
+                    outputPorts.output.disconnect();
+                }
+            }, function( error ) { 
+    
+                // Set error state and send it on over the error port
+                var errorState = stateFactory( vnid, error );
+                vni.errorState( errorState );
+
+                outputPorts.error.output.send( errorState );
+                outputPorts.error.disconnect();
+            }); 
+        }
+     } 
+
      return;
 };
+
+/** 
+ * Check the input to see if we have all data required to run the updater or not. 
+ * This is simply a default updater policy - we may have other policies in the future.
+ * 
+ * @param node node instance of the RDF pipeline component
+ * @param vni virtual node instance whose input states are to be checked.
+ *
+ * @return true if we have received data on all required input port edges
+ */
+function shouldRunUpdater( node, vni ) { 
+
+    var requiredInPorts = requiredInputPorts.call( node );
```

Comment:
This should be checking the *attached* ports -- not the required ports.  The isRequired property only affects the configuration-time requirements -- not the runtime requirements.   Every attached port *must* have data before potentially running the updater, required or not.   But it is a configuration-time error to run with a required port unattached.